Split Transaction proto

tbg opened the pull request:
As noted in #3341, the roachpb.Transaction is one of our large protobufs and
has been persisted in any intent.

With this change, we only store the minimum information necessary for
conflict/intent resolution into each intent, not the whole bulk of the
Transaction proto.

In the process, refactored `MVCCResolveIntent{,Range}` to operate on
our `Intent` proto, which now holds the transaction metadata proto
along with the commit status of that transaction.

I think I've introduced some endless intent resolution loop or something
like that (`kv` fails on occasionally) so I'm going to hunt that down, but this
should still be worth a look
already.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4241)

<!-- Reviewable:end -->

tamird commented:
Reviewed 23 of 23 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[storage/engine/mvcc.go, line 656 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KA2gl07DDCZROaa662p:-KA2gcUejECeFoBZyB96:1618443010) ([raw file](https://github.com/cockroachdb/cockroach/blob/19c1fa8aed2c1b80b58b09a234b4bf8e930c5be2/storage/engine/mvcc.go#L656)):*
Can an intent ever have a non-pending status? why include it on the intent struct at all?

---

*[storage/engine/mvcc.go, line 1332 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KA2gwGa3oBvxxEK8_3p:-KA2gwGa3oBvxxEK8_3q:166874) ([raw file](https://github.com/cockroachdb/cockroach/blob/19c1fa8aed2c1b80b58b09a234b4bf8e930c5be2/storage/engine/mvcc.go#L1332)):*
?

---

*[storage/replica_command.go, line 924 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KA2g2SyL8dQ9sB2reiq:-KA2g2SyL8dQ9sB2reir:-1756720641) ([raw file](https://github.com/cockroachdb/cockroach/blob/19c1fa8aed2c1b80b58b09a234b4bf8e930c5be2/storage/replica_command.go#L924)):*
what happened to the epoch and priority dance above?

---

*[storage/replica_command.go, line 1024 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KA2fgccMd2EFzO0Id2N:-KA2fgccMd2EFzO0Id2O:762517310) ([raw file](https://github.com/cockroachdb/cockroach/blob/19c1fa8aed2c1b80b58b09a234b4bf8e930c5be2/storage/replica_command.go#L1024)):*
s/it's/its/

---

*[storage/replica_command.go, line 1046 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KA2fTZNi8Qr1cMeLxa_:-KA2fTZOffrESOOrDU_w:-1190722286) ([raw file](https://github.com/cockroachdb/cockroach/blob/19c1fa8aed2c1b80b58b09a234b4bf8e930c5be2/storage/replica_command.go#L1046)):*
why not a composite literal?

---

*[storage/replica_command.go, line 1062 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KA2fV_Dhyfxy31E05ei:-KA2fV_Dhyfxy31E05ej:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/19c1fa8aed2c1b80b58b09a234b4bf8e930c5be2/storage/replica_command.go#L1062)):*
ditto

---

*[storage/replica_test.go, line 2143 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KA2gIQhYbm8UDP4qBpc:-KA2gIQixma_SZZaeW7-:-1800428596) ([raw file](https://github.com/cockroachdb/cockroach/blob/19c1fa8aed2c1b80b58b09a234b4bf8e930c5be2/storage/replica_test.go#L2143)):*
is the intetion to resolve this before merging?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241)*
<!-- Sent from Reviewable.io -->

petermattis commented:
`futsch`?

Very nice. I didn't review that thoroughly, though. Mainly concentrated on the proto changes.

---

Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[roachpb/data.proto, line 210 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KA2hrLb7zgW94p4-B-I:-KA2hrLb7zgW94p4-B-J:2023456610) ([raw file](https://github.com/cockroachdb/cockroach/blob/19c1fa8aed2c1b80b58b09a234b4bf8e930c5be2/roachpb/data.proto#L210)):*
`Meta` is a general name. `TxnMeta`? This would make more sense if `s/Transaction/Txn/g`, though that's a heck of a change. 

---

*[roachpb/data.proto, line 217 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KA2jfk3LXviKmk9K6Uk:-KA2jfk4nmGVZ5MTWspq:504302364) ([raw file](https://github.com/cockroachdb/cockroach/blob/19c1fa8aed2c1b80b58b09a234b4bf8e930c5be2/roachpb/data.proto#L217)):*
The intermixing of `txn` and `transaction` in the comments keeps you on your toes.

---

*[roachpb/data.proto, line 222 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KA2jnVSjcXkcjvVgVty:-KA2jnVSjcXkcjvVgVtz:1975056749) ([raw file](https://github.com/cockroachdb/cockroach/blob/19c1fa8aed2c1b80b58b09a234b4bf8e930c5be2/roachpb/data.proto#L222)):*
Wow! So much smaller. Such savings! Curious to see if this has a measurable perf impact. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-:-KA2hUlCF_7DbrDRByYl:1762290153)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 2 of 23 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[roachpb/data.proto, line 210 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KA2hrLb7zgW94p4-B-I:-KA372MtRJsDbVnE168o:-867027520) ([raw file](https://github.com/cockroachdb/cockroach/blob/19c1fa8aed2c1b80b58b09a234b4bf8e930c5be2/roachpb/data.proto#L210)):*
Stutters a bit, but I agree that `Meta` is just really general. It's never used out of context, though. I renamed to `TxnMeta`.

---

*[storage/engine/mvcc.go, line 656 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KA2gl07DDCZROaa662p:-KA34Ra1rCh-Rkm7aIfs:-971904719) ([raw file](https://github.com/cockroachdb/cockroach/blob/19c1fa8aed2c1b80b58b09a234b4bf8e930c5be2/storage/engine/mvcc.go#L656)):*
Yes, if you pass an intent around you have an intention to do something with it: push timestamp, commit, abort.

---

*[storage/engine/mvcc.go, line 1332 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KA2gwGa3oBvxxEK8_3p:-KA34u3ZZ6sTJz9re1AN:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/19c1fa8aed2c1b80b58b09a234b4bf8e930c5be2/storage/engine/mvcc.go#L1332)):*
Done.

---

*[storage/replica_command.go, line 924 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KA2g2SyL8dQ9sB2reiq:-KA34z8Cvd4IoGKNOzp0:-423927194) ([raw file](https://github.com/cockroachdb/cockroach/blob/19c1fa8aed2c1b80b58b09a234b4bf8e930c5be2/storage/replica_command.go#L924)):*
The Pusher now has only the Txn.Meta, so it doesn't know about the Epoch and Priority at all. Previously we were kind of drive-by updating the central proto here since that data was available. Now it's not, and I think that's ok. There are other mechanisms we can explore if fast propagation of the ongoing transaction state into the proto is important.

---

*[storage/replica_test.go, line 2143 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KA2gIQhYbm8UDP4qBpc:-KA36ZkrpSIHoc7bk74j:-1251132345) ([raw file](https://github.com/cockroachdb/cockroach/blob/19c1fa8aed2c1b80b58b09a234b4bf8e930c5be2/storage/replica_test.go#L2143)):*
Planning to look into it tomorrow, but if it turns out to be an invasive change I'll do it separately.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
@petermattis: Not thrilled about this. Ideas?

```
04:24 $ benchstat insert.old insert.new
name                    old time/op    new time/op    delta
Bank_Cockroach-32          208µs ±10%     222µs ±17%    ~           (p=0.247 n=10+10)
Select1_Cockroach-32      73.6µs ± 4%    74.8µs ± 4%  +1.61%         (p=0.043 n=9+10)
Insert1_Cockroach-32       668µs ± 2%     679µs ± 1%  +1.62%         (p=0.002 n=10+8)
Insert10_Cockroach-32     1.21ms ± 1%    1.23ms ± 1%  +2.04%        (p=0.000 n=10+10)
Insert100_Cockroach-32    5.93ms ± 2%    5.95ms ± 2%    ~           (p=0.315 n=10+10)
Update1_Cockroach-32      1.52ms ± 1%    1.55ms ± 5%  +1.77%        (p=0.029 n=10+10)
Update10_Cockroach-32     8.62ms ± 3%    8.90ms ± 2%  +3.23%         (p=0.001 n=10+8)
Update100_Cockroach-32    78.8ms ± 2%    81.8ms ± 3%  +3.84%        (p=0.000 n=10+10)
Delete1_Cockroach-32      1.75ms ± 3%    1.81ms ± 1%  +3.46%        (p=0.000 n=10+10)
Delete10_Cockroach-32     2.77ms ± 2%    2.82ms ± 3%  +1.97%        (p=0.004 n=10+10)
Delete100_Cockroach-32    12.6ms ± 2%    12.7ms ± 4%    ~           (p=0.143 n=10+10)
Scan1_Cockroach-32         355µs ± 2%     360µs ± 4%    ~           (p=0.165 n=10+10)
Scan10_Cockroach-32        425µs ± 3%     431µs ± 2%  +1.40%        (p=0.005 n=10+10)
Scan100_Cockroach-32      1.01ms ± 3%    1.03ms ± 2%    ~           (p=0.247 n=10+10)

name                    old alloc/op   new alloc/op   delta
Bank_Cockroach-32          177kB ± 3%     179kB ± 3%    ~           (p=0.247 n=10+10)
Select1_Cockroach-32      2.15kB ± 1%    2.15kB ± 1%    ~            (p=0.434 n=10+9)
Insert1_Cockroach-32      25.9kB ± 0%    25.9kB ± 0%    ~           (p=0.436 n=10+10)
Insert10_Cockroach-32     83.2kB ± 0%    81.7kB ± 0%  -1.79%         (p=0.000 n=10+8)
Insert100_Cockroach-32     651kB ± 0%     635kB ± 0%  -2.47%          (p=0.000 n=9+9)
Update1_Cockroach-32      43.7kB ± 0%    44.0kB ± 0%  +0.54%         (p=0.000 n=10+9)
Update10_Cockroach-32      283kB ± 0%     284kB ± 0%  +0.34%        (p=0.002 n=10+10)
Update100_Cockroach-32    2.66MB ± 0%    2.67MB ± 0%  +0.25%         (p=0.000 n=9+10)
Delete1_Cockroach-32      41.4kB ± 0%    41.5kB ± 0%  +0.28%        (p=0.030 n=10+10)
Delete10_Cockroach-32      142kB ± 0%     139kB ± 0%  -1.74%        (p=0.000 n=10+10)
Delete100_Cockroach-32    1.05MB ± 0%    1.03MB ± 0%  -2.53%        (p=0.000 n=10+10)
Scan1_Cockroach-32        11.3kB ± 1%    11.4kB ± 0%  +0.82%         (p=0.000 n=10+9)
Scan10_Cockroach-32       17.7kB ± 0%    17.8kB ± 0%  +0.64%        (p=0.000 n=10+10)
Scan100_Cockroach-32      70.3kB ± 0%    70.4kB ± 0%  +0.14%        (p=0.014 n=10+10)

name                    old allocs/op  new allocs/op  delta
Bank_Cockroach-32          1.04k ± 1%     1.07k ± 4%    ~            (p=0.177 n=8+10)
Select1_Cockroach-32        49.0 ± 0%      49.0 ± 0%    ~     (all samples are equal)
Insert1_Cockroach-32         373 ± 0%       371 ± 0%  -0.54%         (p=0.000 n=10+9)
Insert10_Cockroach-32        877 ± 0%       857 ± 0%  -2.33%          (p=0.000 n=9+8)
Insert100_Cockroach-32     5.63k ± 0%     5.43k ± 0%  -3.58%         (p=0.000 n=10+9)
Update1_Cockroach-32         826 ± 0%       824 ± 0%  -0.24%        (p=0.000 n=10+10)
Update10_Cockroach-32      5.68k ± 0%     5.66k ± 0%  -0.34%        (p=0.000 n=10+10)
Update100_Cockroach-32     54.0k ± 0%     53.8k ± 0%  -0.36%        (p=0.000 n=10+10)
Delete1_Cockroach-32         766 ± 0%       764 ± 0%  -0.26%        (p=0.000 n=10+10)
Delete10_Cockroach-32      1.66k ± 0%     1.64k ± 0%  -1.24%         (p=0.000 n=9+10)
Delete100_Cockroach-32     9.95k ± 0%     9.75k ± 0%  -2.01%         (p=0.000 n=10+9)
Scan1_Cockroach-32           224 ± 0%       224 ± 0%    ~           (p=1.000 n=10+10)
Scan10_Cockroach-32          326 ± 0%       327 ± 0%  +0.21%         (p=0.008 n=10+8)
Scan100_Cockroach-32       1.24k ± 0%     1.24k ± 0%    ~           (p=1.000 n=10+10)
```

tschottdorf commented:
FWIW, this is on GCE, on a local run the numbers went slightly the other way.

tamird commented:
Reviewed 21 of 21 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[roachpb/data.go, line 842 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KA6fRbjJaonzjccHS3j:-KA6fRbk1zzsYia_SWKk:-758243026) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3ba3bc3e84945ce0bafc306ee1a1f248d73d0b5/roachpb/data.go#L842)):*
Make these assignments on three lines.

---

*[roachpb/data.proto, line 218 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KA6e2UaEUKhRPKu-1T9:-KA6e2UaEUKhRPKu-1TA:-1978816902) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3ba3bc3e84945ce0bafc306ee1a1f248d73d0b5/roachpb/data.proto#L218)):*
Might as well renumber these fields while you're moving things around.

---

*[storage/engine/mvcc.proto, line 26 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KA6gAJuh9IIdDWG8DA7:-KA6gAJuh9IIdDWG8DA8:-1465955028) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3ba3bc3e84945ce0bafc306ee1a1f248d73d0b5/storage/engine/mvcc.proto#L26)):*
Maybe it's only confusing in the context of this PR, but I find all the variables named `[Tt]xn` of type `TxnMeta` to look weird. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[roachpb/data.go, line 842 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KA6fRbjJaonzjccHS3j:-KA6igcvSNwo5KRrItAH:-283318394) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3ba3bc3e84945ce0bafc306ee1a1f248d73d0b5/roachpb/data.go#L842)):*
or use a composite literal

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[roachpb/data.go, line 842 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KA6fRbjJaonzjccHS3j:-KA8nd5MzV6UvDYjcN1P:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3ba3bc3e84945ce0bafc306ee1a1f248d73d0b5/roachpb/data.go#L842)):*
Done.

---

*[roachpb/data.proto, line 218 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KA6e2UaEUKhRPKu-1T9:-KA8o4d7SmM3zqG2O4R0:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3ba3bc3e84945ce0bafc306ee1a1f248d73d0b5/roachpb/data.proto#L218)):*
Done.

---

*[storage/engine/mvcc.proto, line 26 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KA6gAJuh9IIdDWG8DA7:-KA8oCeX7ClGQs5HJw4f:1870426325) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3ba3bc3e84945ce0bafc306ee1a1f248d73d0b5/storage/engine/mvcc.proto#L26)):*
Suggestions?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Ok, tests are passing. Should be ready for a final look.

tamird commented:
Reviewed 8 of 8 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[storage/engine/mvcc.go, line 1360 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KAAdjBsSKuYlYIhvIkJ:-KAAdjBsSKuYlYIhvIkK:-854914614) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f5b264979642109b0eca70f210baa3c54aed798/storage/engine/mvcc.go#L1360)):*
isn't this error message the inverse of the condition you're checking above? meta's epoch is greater than the intent's,

---

*[storage/engine/mvcc.go, line 1363 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KAAeEAoTQBevxJuRpKW:-KAAeEAoTQBevxJuRpKX:205382865) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f5b264979642109b0eca70f210baa3c54aed798/storage/engine/mvcc.go#L1363)):*
again, isn't this the inverse of the code below? it seems to be allowing a push from a future epoch. are the variable names here not what i think they are?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[storage/engine/mvcc.go, line 1354 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KABj-KACqC9qT9bTL6O:-KABj-KACqC9qT9bTL6P:573354754) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f5b264979642109b0eca70f210baa3c54aed798/storage/engine/mvcc.go#L1354)):*
Is there some test failure that prompted this change? Is the change necessary regardless of the Transaction proto split or is it only necessary because we are now dealing with incomplete information in some cases?

---

*[storage/engine/mvcc.go, line 1377 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KABdhnzfnKHNnH3LS52:-KABdho-cLlM4HPcnbU4:-1251802241) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f5b264979642109b0eca70f210baa3c54aed798/storage/engine/mvcc.go#L1377)):*
If B successfully pushed, then the write intent at epoch 2 would have a higher timestamp and so B wouldn't see it on retry, right?

---

*[storage/engine/mvcc.proto, line 26 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KA6gAJuh9IIdDWG8DA7:-KAB_W2FCnorRuSWUW3F:953960692) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3ba3bc3e84945ce0bafc306ee1a1f248d73d0b5/storage/engine/mvcc.proto#L26)):*
My suggestion would be to use `TxnMeta` instead of `Txn` as the default name of any variable of type `TxnMeta`. Not sure if it's worth the churn, though (maybe just for exported fields?)

---

*[storage/replica_command.go, line 917 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KABjMHcad4g7-ogo0kg:-KABjMHcad4g7-ogo0kh:1803397771) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f5b264979642109b0eca70f210baa3c54aed798/storage/replica_command.go#L917)):*
Does PusheeTxn need to be a full Transaction or could it be a TxnMeta+Status like we have in the intent? 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[storage/engine/mvcc.go, line 1354 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KABj-KACqC9qT9bTL6O:-KABqlxKWQ1Vf0boeHN6:-2085231778) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f5b264979642109b0eca70f210baa3c54aed798/storage/engine/mvcc.go#L1354)):*
I'm assuming that by "this change" you mean the definition of `pushed` below (`commit` hasn't changed). The reason for the change is in the table below, this was tickled in the kv correctness tests. The reason this shows up is a small loss of fidelity I just fixed: previously, a Push would update the persisted Epoch of the Pushee (if larger), which means that a subsequent reiteration of the conflict would update the returned Pushee, and so the intent resolution would have the updated Epoch as well, making the old code work the second time around. Updating the below comment.

---

*[storage/engine/mvcc.go, line 1360 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KAAdjBsSKuYlYIhvIkJ:-KABy16LeLBIJ4ZLXvfJ:-1187638600) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f5b264979642109b0eca70f210baa3c54aed798/storage/engine/mvcc.go#L1360)):*
`intent` is what comes from the intent (which is what we're trying to commit), meta.Txn is what the actual intent has.

---

*[storage/engine/mvcc.go, line 1363 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KAAeEAoTQBevxJuRpKW:-KAByCnrU7U_HzdzdDS0:-1135163865) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f5b264979642109b0eca70f210baa3c54aed798/storage/engine/mvcc.go#L1363)):*
`meta.Txn.Epoch` is what's here, `intent` is what the Pusher found (a while back).

---

*[storage/engine/mvcc.go, line 1377 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KABdhnzfnKHNnH3LS52:-KAByTS8XXI5En-Zd7CM:-126245989) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f5b264979642109b0eca70f210baa3c54aed798/storage/engine/mvcc.go#L1377)):*
assuming that A restarted with a timestamp that's still below B, it would. But the second iteration would actually propagate the Epoch to the ResolveIntent, so it's fine. I just fixed that and updated comment.

---

*[storage/engine/mvcc.proto, line 26 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KA6gAJuh9IIdDWG8DA7:-KAC-bAwA5rRiml9dRgj:818627532) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3ba3bc3e84945ce0bafc306ee1a1f248d73d0b5/storage/engine/mvcc.proto#L26)):*
I'd like to avoid the churn for now. Easy enough to do (and more fun to review) in isolation.

---

*[storage/replica_command.go, line 917 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KABjMHcad4g7-ogo0kg:-KAC05rmGOzR6K2K5nhm:-1305565536) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f5b264979642109b0eca70f210baa3c54aed798/storage/replica_command.go#L917)):*
Not unlikely. Added a TODO; should be easy enough for someone to pick up separately - minor plumbing will be required and may not be worth it.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241)*
<!-- Sent from Reviewable.io -->

tbg committed:
Split Transaction proto

As noted in #3341, the roachpb.Transaction is one of our large protobufs and
has been persisted in any intent.

With this change, we only store the minimum information necessary for
conflict/intent resolution into each intent, not the whole bulk of the
Transaction proto.

In the process, refactored `MVCCResolveIntent{,Range}` to operate on
our `Intent` proto, which now holds the transaction metadata proto
along with the commit status of that transaction.

Tickled and a bug in MVCCResolveIntent which lead to an infinite retry
loop: a successful Push's ResolveIntent would under special circumstances
not actually move the intent forward in time.

tamird commented:
Review status: 20 of 24 files reviewed at latest revision, 6 unresolved discussions.

---

*[storage/engine/mvcc.go, line 1363 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KAAeEAoTQBevxJuRpKW:-KAC1I4hslpAVXj9-Awi:1194498714) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f5b264979642109b0eca70f210baa3c54aed798/storage/engine/mvcc.go#L1363)):*
I seee. Confusing naming for the uninitiated.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 20 of 24 files reviewed at latest revision, 6 unresolved discussions.

---

*[storage/engine/mvcc.go, line 1363 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-KAAeEAoTQBevxJuRpKW:-KAC2F-Br2Md5sdHwQav:511744509) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f5b264979642109b0eca70f210baa3c54aed798/storage/engine/mvcc.go#L1363)):*
As always, I'm inviting comments on names (and LGTMs).

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

---

Review status: 19 of 24 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4241#-:-KACCZddBXGX7uzTF1rb:676834233)*
<!-- Sent from Reviewable.io -->
